### PR TITLE
HPCC-19216 Dynamically add <Properties> to ESPService and ESPBinding sections

### PR DESCRIPTION
**Pull request by kenrowland**
Add support to dynamically add <Properties> sections to
ESPService and ESPBinding sections in the environment

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by kenrowland** [comment]
@rpastrana @afishbeck Please review

This PR adds event processing for element creating that inserts XML (properties and authentication features) into dependent sections, namely ESPService and ESPBinding.

It also adds dependent attribute values. This handles default http and https port numbers for ESPBindings.

File: configuration/config2/EnvironmentEvents.cpp
Comment (afishbeck) [inline configuration/config2/EnvironmentEvents.cpp]: Should cases like these be "unique_ptr" rather than shared_ptr?  If so, there are probably many cases throughout.

File: configuration/config2/EnvironmentLoader.hpp
Comment (afishbeck) [inline configuration/config2/EnvironmentLoader.hpp]: Not really a fan of keeping commented out code... are these functions going to be readded at some point?

File: configuration/config2/SchemaItem.hpp
Comment (afishbeck) [inline configuration/config2/SchemaItem.hpp]: I understand the implementation "walks" the parent nodes, but conceptually this doesn't seem like a "find" function,.. more like a "getSchemaRoot()" or similiar name.

File: configuration/config2/EnvironmentEvents.hpp
Comment (afishbeck) [inline configuration/config2/EnvironmentEvents.hpp]: Trivial, but lots of extra space.

File: configuration/config2/EnvironmentEvents.hpp
Comment (afishbeck) [inline configuration/config2/EnvironmentEvents.hpp]: Seems like you are relying on the order these methods may be called to change the state.  i.e. what if I did:

setMatchAttributeName("match");
setItemAttributeName("item");

?

Unless that's intentional, then in case setMatchAttributeName was already called I would do something like:
void setItemAttributeName(const std::string &name) 
{ 
m_itemAttribute = name; 
if (m_matchAttribute.empty())
    m_matchAttribute = name;
}

File: configuration/config2/EnvironmentMgr.cpp
Comment (afishbeck) [inline configuration/config2/EnvironmentMgr.cpp]: this block of comment code is particularly ugly.

File: configuration/config2/EnvironmentMgr.cpp
Comment (afishbeck) [inline configuration/config2/EnvironmentMgr.cpp]: Wonder about the name of these events being "create".  Can you explain the meaning of create... when the comment above makes it sound like an "added".

File: configuration/config2/InsertableItem.cpp
Comment (afishbeck) [inline configuration/config2/InsertableItem.cpp]: Again lots of variables that I think probably could be unique_ptr.

File: configuration/config2/InsertableItem.hpp
Comment (afishbeck) [inline configuration/config2/InsertableItem.hpp]: Do you use struct just to avoid a "public:" declaration?

File: esp/services/ws_config2/ws_config2_binding.hpp
Comment (afishbeck) [inline esp/services/ws_config2/ws_config2_binding.hpp]: missing carriage return.

**Review by afishbeck — COMMENTED** [review]
Generally looks ok.  Some comments.  There may be places where you should be using the "override" declaration when overriding virtual functions.  Didn't take the time to track down specific ones.

**Comment by afishbeck** [comment]
@stuartort someone else should also review.  Especially details of c++ and std library usage.

File: configuration/config2/XMLEnvironmentLoader.cpp
Comment (rpastrana) [inline configuration/config2/XMLEnvironmentLoader.cpp]: didn't notice any member vars, is this method static by nature/should it be declared as such

File: esp/scm/ws_config2.ecm
Comment (rpastrana) [inline esp/scm/ws_config2.ecm]: spelling

File: esp/scm/ws_config2.ecm
Comment (rpastrana) [inline esp/scm/ws_config2.ecm]: It might be more common in hpcc code to call this FetchNodesRequest

File: esp/scm/ws_config2.ecm
Comment (rpastrana) [inline esp/scm/ws_config2.ecm]: FetchNodes?

File: initfiles/componentfiles/config2xml/environment.xml
Comment (rpastrana) [inline initfiles/componentfiles/config2xml/environment.xml]: what's this file used for? is it here for testing?

File: initfiles/componentfiles/config2xml/environment190.xml
Comment (rpastrana) [inline initfiles/componentfiles/config2xml/environment190.xml]: This file seems out of place in this folder, it prob shouldn't be checked in at all

**Review by rpastrana — CHANGES_REQUESTED** [review]
@kenrowland several comments/questions. Noticed commented out code throughout. Remove 190env file

File: configuration/config2/EnvironmentEvents.cpp
Comment (kenrowland) [reply configuration/config2/EnvironmentEvents.cpp]: In most cases the called method returns a shared_ptr since because the object already exists and most likely has a reference. Converting to a unique_ptr in these cases would not follow the architecture since unique ownership is not the intent. I will, however, keep an eye out for situations where a unique_ptr would suffice, such as where the method creates the object and persistence is not necessary.

File: configuration/config2/EnvironmentEvents.hpp
Comment (afishbeck) [reply configuration/config2/EnvironmentEvents.hpp]: Or an EnvironmentCreateEvent?

**Comment by afishbeck** [comment]
@kenrowland if you wait until the end of code review to rebase it will make things much easier for the reviewers.

File: configuration/config2/EnvironmentLoader.hpp
Comment (kenrowland) [reply configuration/config2/EnvironmentLoader.hpp]: Methods were moved during some refactoring. I neglected to remove the commented lines. Lines now removed.

File: configuration/config2/SchemaItem.hpp
Comment (kenrowland) [reply configuration/config2/SchemaItem.hpp]: True, it is a get, find was carried over because it was written while I was implementing the find functionality. I changed it to getSchemaRoot. Similar change made to finding the root node in the environment.

File: configuration/config2/InsertableItem.hpp
Comment (kenrowland) [reply configuration/config2/InsertableItem.hpp]: I suppose yes. It has been a style I have used that when a small simply all public class is necessary, I just use a struct to indicate that it's all public. I also use it when instantiated objects are treated more like data, such as items in a vector.

File: configuration/config2/SchemaItem.cpp
Comment (kenrowland) [reply configuration/config2/SchemaItem.cpp]: Comment no longer valid. Removed

File: configuration/config2/Utils.cpp
Comment (kenrowland) [reply configuration/config2/Utils.cpp]: I put this code in a common module for use by config2 manager. It's a static function that is used by a number of methods. Being located in the config2 directory I figured was enough to indicate it's for config2. If this should be moved to a system wide location, please tell me where. I am hesitant to move it unless it becomes part of a system lib because it would only be compiled and used by config2.

File: configuration/config2/XMLEnvironmentLoader.cpp
Comment (kenrowland) [reply configuration/config2/XMLEnvironmentLoader.cpp]: The intent is that you use this method when you have an instance of the object, not to be able to call it otherwise. In the future there could be use of class member variables.

File: esp/scm/ws_config2.ecm
Comment (kenrowland) [reply esp/scm/ws_config2.ecm]: Unless there is a compelling reason, I will leave this as "find". A change to fetch will percolate all the way through the config manager.

File: esp/scm/ws_config2.ecm
Comment (kenrowland) [reply esp/scm/ws_config2.ecm]: Fixed.

File: esp/scm/ws_config2.ecm
Comment (kenrowland) [reply esp/scm/ws_config2.ecm]: See previous comment. Leaving as find to keep same paradigm throughout all of config manager.

File: esp/services/ws_config2/ws_config2_binding.hpp
Comment (kenrowland) [reply esp/services/ws_config2/ws_config2_binding.hpp]: Fixed.

**Review by kenrowland — COMMENTED** [review]
Comments addressed. New commit will follow shortly.

**Comment by kenrowland** [comment]
@afishbeck , @rpastrana 
Changes have been made. Please review latest commit.

Thanks.

File: configuration/config2/EnvironmentEventHandlers.cpp
Comment (rpastrana) [inline configuration/config2/EnvironmentEventHandlers.cpp]: this cr symbol looks odd

**Review by rpastrana — CHANGES_REQUESTED** [review]
@kenrowland one minor issue

**Comment by rpastrana** [comment]
Sending  back to @kenrowland

File: configuration/config2/EnvironmentEventHandlers.cpp
Comment (kenrowland) [reply configuration/config2/EnvironmentEventHandlers.cpp]: Fixed. I believe the issue is when I create a file in Windows, and push it. I think the character is different between Linux and Windows.

**Comment by richardkchapman** [comment]
@rpastrana Are you ok with this now? If so, will need squashing

**Comment by kenrowland** [comment]
@rpastrana Renamed APIs to FetchXXX. Please review and approve.
@afishbeck Please approve if no further comments

**Review by afishbeck — APPROVED** [review]
@kenrowland looks ok

**Comment by kenrowland** [comment]
@rpastrana If no further comments, please approve. Thanks

**Comment by rpastrana** [comment]
@kenrowland latest commit looks reasonable, you can squash your commits.

**Comment by kenrowland** [comment]
@richardkchapman Please merge

**Pull request merged**
